### PR TITLE
PYTHON-2773 Mockupdb test failures

### PR DESCRIPTION
**Pull request by juliusgeo**

### CODE REVIEW

File: test/mockupdb/test_list_indexes.py [hunk 1]
```diff
@@ -1,78 +0,0 @@
...
-                                      namespace='test.collection',
-                                      cursor_id=123)
-
-            request.reply([{'name': 'index_1'}], cursor_id=0)
-
-        self.assertEqual([{'name': 'index_0'}, {'name': 'index_1'}], indexes())
-        for index_info in indexes():
-            self.assertIsInstance(index_info, SON)
-
-    def test_list_indexes_command(self):
```
Comment (ShaneHarvey) [inline test/mockupdb/test_list_indexes.py:50]: This test still seems relevant, and it appears to be passing:
```
 [2021/11/16 02:54:30.734] test_list_indexes_command (tests.test_list_indexes.TestListIndexes) ... ok
```

Let's keep it.

File: test/mockupdb/test_mongos_command_read_mode.py [hunk 2]
```diff
@@ -39,18 +39,14 @@ def test_aggregate(self):
...
             read_preference=ReadPreference.SECONDARY)
 
         with going(secondary_collection.aggregate, []):
-            command = server.receives(
-                {'$query': SON([('aggregate', 'collection'),
-                                ('pipeline', []),
-                                ('cursor', {})]),
-                 '$readPreference': {'mode': 'secondary'}})
+
+            command = server.receives(OpMsg({"aggregate": "collection", "pipeline": []}))
```
Comment (ShaneHarvey) [inline test/mockupdb/test_mongos_command_read_mode.py:49]: `'$readPreference': {'mode': 'secondary'}})` is an important part of this test, let's add it back.

File: test/mockupdb/test_mongos_command_read_mode.py [hunk 3]
```diff
@@ -61,37 +57,29 @@ def test(self):
...
                                     tag_sets=None)
 
         client = MongoClient(server.uri, read_preference=pref)
         self.addCleanup(client.close)
-        with going(operation.function, client) as future:
-            request = server.receive()
-            request.reply(operation.reply)
+        future = go(operation.function, client)
+        request = server.receive()
+        request.reply(operation.reply)
```
Comment (ShaneHarvey) [inline test/mockupdb/test_mongos_command_read_mode.py:69]: What's this change for?

File: test/mockupdb/test_op_msg_read_preference.py [hunk 4]
```diff
@@ -159,11 +159,10 @@ def test(self):
             expected_pref = pref
         else:
             self.fail('unrecognized op_type %r' % operation.op_type)
-
         # For single mongod we send primaryPreferred instead of primary.
-        if expected_pref == ReadPreference.PRIMARY and self.single_mongod:
+        if (expected_pref == ReadPreference.PRIMARY and self.single_mongod
+                and operation.name != "command"):
```
Comment (ShaneHarvey) [inline test/mockupdb/test_op_msg_read_preference.py:164]: Why is this needed?

File: test/mockupdb/test_projection.py [hunk 5]
```diff
@@ -15,7 +15,7 @@
 """PyMongo shouldn't append projection fields to "find" command, PYTHON-1479."""
```
Comment (ShaneHarvey) [inline test/mockupdb/test_projection.py:15]: We can remove this file. This test only made sense in OP_QUERY.

File: test/mockupdb/test_query_read_pref_sharded.py [hunk 6]
```diff
@@ -30,7 +30,7 @@ class TestQueryAndReadModeSharded(unittest.TestCase):
     def test_query_and_read_mode_sharded_op_query(self):
```
Comment (ShaneHarvey) [inline test/mockupdb/test_query_read_pref_sharded.py:30]: This test is for op-query so we can remove it. There's already an op_msg test below.

File: test/mockupdb/test_starting_from_overflow.py [hunk 7]
```diff
@@ -14,7 +14,7 @@
 
 """Test that PyMongo ignores the startingFrom field, PYTHON-945."""
```
Comment (ShaneHarvey) [inline test/mockupdb/test_starting_from_overflow.py:15]: This test only makes sense for OP_REPLY+OP_QUERY. Let's remove this file.

File: test/mockupdb/test_query_read_pref_sharded.py [hunk 8]
```diff
@@ -53,19 +53,12 @@ def test_query_and_read_mode_sharded_op_query(self):
...
-                                 ('$readPreference', mode.document)])))
 
-                    request.replies({'cursor': {'id': 0, 'firstBatch': [{}]}})
+                command = OpMsg(find_command)
+
+                go(next, cursor)
+                request = server.receives(command)
+                request.replies({'cursor': {'id': 0, 'firstBatch': [{}]}})
 
     @unittest.skipUnless(version_tuple >= (3, 7), "requires PyMongo 3.7")
```
Comment (ShaneHarvey) [inline test/mockupdb/test_query_read_pref_sharded.py:63]: Please search for `version_tuple` in test/mockupdb/ and remove all code that uses pymongo's version and remove any dead code. Now that the tests live in pymongo itself we always know the version.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
A few suggestions.

File: test/mockupdb/operations.py [hunk 9]
```diff
@@ -76,20 +76,6 @@
         reply={'cursor': {'id': 0, 'firstBatch': []}},
         op_type='may-use-secondary',
         not_master=not_master_reply_to_command),
```
Comment (ShaneHarvey) [inline test/mockupdb/operations.py:78]: not_master_reply_to_command and not_master_reply_to_query need to be OpMsgReply replies.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: test/mockupdb/operations.py (hunk 9 above)
Comment (juliusgeo) [reply test/mockupdb/operations.py:78]: Done.

File: test/mockupdb/test_op_msg_read_preference.py (hunk 4 above)
Comment (ShaneHarvey) [reply test/mockupdb/test_op_msg_read_preference.py:162]: That might be a pymongo bug. Can you open a new ticket to investigate?

File: test/mockupdb/test_mongos_command_read_mode.py (hunk 3 above)
Comment (ShaneHarvey) [reply test/mockupdb/test_mongos_command_read_mode.py:71]: I'm still confused. The docs you quoted say that using "going" is more succinct and we also don't use the return value in this test, so why would we change the code to use the more verbose option ("go")?

I think we should change this to use going and remove the explicit "future":
```
        with going(operation.function, client):
            request = server.receive()
            request.reply(operation.reply)
```

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: test/mockupdb/test_mongos_command_read_mode.py (hunk 3 above)
Comment (juliusgeo) [reply test/mockupdb/test_mongos_command_read_mode.py]: Done.

File: test/mockupdb/test_op_msg_read_preference.py (hunk 4 above)
Comment (juliusgeo) [reply test/mockupdb/test_op_msg_read_preference.py:162]: Done.

**Review by ShaneHarvey — APPROVED** [review]
LGTM
